File: backend/tests/validation_tasks/arch06/session_store.py
"""Agent local SQLite session cache (stdlib sqlite3).

Mirrors the production ``agent/src/main/session_store.ts`` (Feature 2.2.3) and
SDD §7.7 ``LocalSessionCache`` — only the fields needed for SYNC.
"""
from __future__ import annotations

import sqlite3
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Optional


SCHEMA = """
CREATE TABLE IF NOT EXISTS local_sessions (
    session_id            TEXT PRIMARY KEY,
    seat_id               TEXT,
    started_at            TEXT,
    local_elapsed_seconds REAL,
    disconnect_at         TEXT,
    reconnect_at          TEXT,
    disconnect_count      INTEGER DEFAULT 0,
    is_synced             INTEGER DEFAULT 0,
    updated_at            TEXT
);
"""


@dataclass
class LocalSession:
    session_id: str
    seat_id: str
    started_at: str            # ISO8601
    local_elapsed_seconds: float
    disconnect_at: Optional[str]
    reconnect_at: Optional[str]
    disconnect_count: int
    is_synced: bool
    updated_at: str


class SessionStore:
    """Thin synchronous wrapper over a per-agent SQLite file."""

    def __init__(self, path: str) -> None:
        self.conn = sqlite3.connect(path)
        self.conn.row_factory = sqlite3.Row
        self.conn.executescript(SCHEMA)
        self.conn.commit()

    def close(self) -> None:
        self.conn.close()

    def persist_session(
        self,
        session_id: str,
        seat_id: str,
        started_at: str,
        local_elapsed_seconds: float = 0.0,
    ) -> None:
        self.conn.execute(
            """INSERT INTO local_sessions
               (session_id, seat_id, started_at, local_elapsed_seconds,
                disconnect_at, reconnect_at, disconnect_count, is_synced, updated_at)
               VALUES (?, ?, ?, ?, NULL, NULL, 0, 0, ?)
               ON CONFLICT(session_id) DO UPDATE SET
                 seat_id=excluded.seat_id,
                 started_at=excluded.started_at,
                 local_elapsed_seconds=excluded.local_elapsed_seconds,
                 updated_at=excluded.updated_at""",
            (session_id, seat_id, started_at, local_elapsed_seconds, _now_iso()),
        )
        self.conn.commit()

    def update_elapsed(self, session_id: str, seconds: float) -> None:
        """The 10s-cadence write (FR-AGENT-008)."""
        self.conn.execute(
            """UPDATE local_sessions
               SET local_elapsed_seconds = ?, updated_at = ?
               WHERE session_id = ?""",
            (seconds, _now_iso(), session_id),
        )
        self.conn.commit()

    def mark_disconnect(self, session_id: str, disconnect_at: str) -> None:
        """The disconnect flush (SDD §7.7 step 1) — bounds ALE staleness at
        reconnect to ~0."""
        self.conn.execute(
            """UPDATE local_sessions
               SET disconnect_at = ?, disconnect_count = disconnect_count + 1,
                   is_synced = 0, updated_at = ?
               WHERE session_id = ?""",
            (disconnect_at, _now_iso(), session_id),
        )
        self.conn.commit()

    def get_for_sync(self, session_id: str) -> Optional[LocalSession]:
        row = self.conn.execute(
            "SELECT * FROM local_sessions WHERE session_id = ?",
            (session_id,),
        ).fetchone()
        if row is None:
            return None
        return _row_to_session(row)

    def mark_synced(self, session_id: str, reconnect_at: str) -> None:
        self.conn.execute(
            """UPDATE local_sessions
               SET is_synced = 1, reconnect_at = ?, updated_at = ?
               WHERE session_id = ?""",
            (reconnect_at, _now_iso(), session_id),
        )
        self.conn.commit()


def _row_to_session(row: sqlite3.Row) -> LocalSession:
    return LocalSession(
        session_id=row["session_id"],
        seat_id=row["seat_id"],
        started_at=row["started_at"],
        local_elapsed_seconds=row["local_elapsed_seconds"],
        disconnect_at=row["disconnect_at"],
        reconnect_at=row["reconnect_at"],
        disconnect_count=row["disconnect_count"],
        is_synced=bool(row["is_synced"]),
        updated_at=row["updated_at"],
    )


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()

File: agent/src/main/storage/session_store.ts
/**
 * Better-sqlite3-based session store for the Arcade agent.
 *
 * Provides synchronous SQLite persistence so session state survives
 * crashes and reconnections. Used by the WebSocket client to cache
 * session data before sending a SYNC message after reconnection.
 */

import Database from 'better-sqlite3';
import type { LocalSessionCache, SessionStore, LocalSessionStatus } from './types.js';

export class BetterSqliteSessionStore implements SessionStore {
  private db: Database.Database | null = null;

  constructor(private readonly dbPath: string = ':memory:') {}

  /** Initialise the database and create the schema. */
  init(): void {
    this.db = new Database(this.dbPath);
    this.db.exec(`
      CREATE TABLE IF NOT EXISTS sessions (
        session_id TEXT PRIMARY KEY,
        seat_id TEXT NOT NULL,
        started_at TEXT NOT NULL,
        local_elapsed_seconds INTEGER NOT NULL DEFAULT 0,
        disconnect_at TEXT,
        is_synced INTEGER NOT NULL DEFAULT 0,
        status TEXT NOT NULL DEFAULT 'ACTIVE',
        updated_at TEXT NOT NULL
      );
      CREATE INDEX IF NOT EXISTS idx_sessions_is_synced ON sessions(is_synced);
    `);
  }

  /** Persist a new session when HIDE_OVERLAY is received. */
  persistSession(sessionId: string, seatId: string, startedAt: string): void {
    if (!this.db) throw new Error('DB not initialised. Call init() first.');
    const sql = this.db.prepare(`
      INSERT INTO sessions (session_id, seat_id, started_at, local_elapsed_seconds, disconnect_at, is_synced, status, updated_at)
      VALUES (?, ?, ?, 0, ?, 0, 'ACTIVE', ?)
      ON CONFLICT(session_id) DO UPDATE SET
        seat_id = excluded.seat_id,
        started_at = excluded.started_at,
        local_elapsed_seconds = excluded.local_elapsed_seconds,
        disconnect_at = excluded.disconnect_at,
        is_synced = excluded.is_synced,
        status = excluded.status,
        updated_at = excluded.updated_at;
    `);
    sql.run(sessionId, seatId, startedAt, null, new Date().toISOString());
  }

  /** Update elapsed time (called every 10s during active session). */
  updateElapsed(sessionId: string, elapsedSeconds: number): void {
    if (!this.db) throw new Error('DB not initialised. Call init() first.');
    const sql = this.db.prepare(`
      UPDATE sessions
      SET local_elapsed_seconds = ?, updated_at = ?
      WHERE session_id = ?;
    `);
    sql.run(elapsedSeconds, new Date().toISOString(), sessionId);
  }

  /** Record disconnect timestamp on WS close. */
  markDisconnect(sessionId: string, disconnectAt: string): void {
    if (!this.db) throw new Error('DB not initialised. Call init() first.');
    const sql = this.db.prepare(`
      UPDATE sessions
      SET disconnect_at = ?, updated_at = ?
      WHERE session_id = ?;
    `);
    sql.run(disconnectAt, new Date().toISOString(), sessionId);
  }

  /** Retrieve the first unsynced session for the SYNC payload. */
  getUnsyncedSession(): LocalSessionCache | null {
    if (!this.db) throw new Error('DB not initialised. Call init() first.');
    const sql = this.db.prepare(`
      SELECT * FROM sessions WHERE is_synced = 0 LIMIT 1;
    `);
    const row = sql.get() as Record<string, unknown> | undefined;
    if (!row) return null;
    return this.mapRow(row);
  }

  /** Mark a session as synced after a successful SYNC ack from the server. */
  markSynced(sessionId: string): void {
    if (!this.db) throw new Error('DB not initialised. Call init() first.');
    const sql = this.db.prepare(`
      UPDATE sessions
      SET is_synced = 1, updated_at = ?
      WHERE session_id = ?;
    `);
    sql.run(new Date().toISOString(), sessionId);
  }

  /** Clear local cache when session ends (SHOW_OVERLAY). */
  clearSession(sessionId: string): void {
    if (!this.db) throw new Error('DB not initialised. Call init() first.');
    const sql = this.db.prepare(`DELETE FROM sessions WHERE session_id = ?;`);
    sql.run(sessionId);
  }

  /** Close the database connection. */
  close(): void {
    if (this.db) {
      this.db.close();
      this.db = null;
    }
  }

  private mapRow(row: Record<string, unknown>): LocalSessionCache {
    return {
      session_id: String(row.session_id),
      seat_id: String(row.seat_id),
      started_at: String(row.started_at),
      local_elapsed_seconds: Number(row.local_elapsed_seconds || 0),
      disconnect_at: row.disconnect_at ? String(row.disconnect_at) : null,
      is_synced: row.is_synced === 1 || row.is_synced === true,
      status: String(row.status) as LocalSessionStatus,
      updated_at: String(row.updated_at),
    };
  }
}
